From: Alessandro Ghedini <al3xbio@gmail.com>
Date: Sun, 1 May 2011 20:56:49 +0200
Subject: build dynamic library with proper name and symlinks

Build the library using the libname.so.X.Y.Z namescheme, and generate
libname.so.X and libname.so symbolic links.
---
 Makefile |   13 +++++++++----
 1 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index b4c6b68..b6b8f4f 100644
--- a/Makefile
+++ b/Makefile
@@ -5,6 +5,8 @@
 OBJ=net.o hiredis.o sds.o async.o
 BINS=hiredis-example hiredis-test
 LIBNAME=libhiredis
+VERSION=$(shell dpkg-parsechangelog | sed -rne 's,^Version: ([^-]+).*,\1,p')
+MAJVERSION=$(shell echo $(VERSION) | sed -rne 's,([^.]+).*,\1,p')
 
 uname_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
 OPTIMIZATION?=-O3
@@ -36,8 +38,9 @@ else
   LDFLAGS?=-L.
   DYLIBSUFFIX=so
   STLIBSUFFIX=a
-  DYLIBNAME?=$(LIBNAME).$(DYLIBSUFFIX)
-  DYLIB_MAKE_CMD?=gcc -shared -Wl,-soname,$(DYLIBNAME) -o $(DYLIBNAME)
+  DYLIBNAME?=$(LIBNAME).$(DYLIBSUFFIX).$(VERSION)
+  DYLIBSONAME=$(LIBNAME).$(DYLIBSUFFIX).$(MAJVERSION)
+  DYLIB_MAKE_CMD?=gcc -shared -Wl,-soname,$(DYLIBSONAME) -o $(DYLIBNAME)
   STLIBNAME?=$(LIBNAME).$(STLIBSUFFIX)
   STLIB_MAKE_CMD?=ar rcs $(STLIBNAME)
 endif
@@ -65,6 +68,8 @@ test.o: test.c hiredis.h
 
 $(DYLIBNAME): $(OBJ)
 	$(DYLIB_MAKE_CMD) $(OBJ)
+	ln -s $(DYLIBNAME) $(DYLIBSONAME)
+	ln -s $(DYLIBSONAME) $(LIBNAME).$(DYLIBSUFFIX)
 
 $(STLIBNAME): $(OBJ)
 	$(STLIB_MAKE_CMD) $(OBJ)
@@ -98,7 +103,7 @@ test: hiredis-test
 	$(CC) -std=c99 -pedantic -c $(CFLAGS) $(OBJARCH) $(DEBUG) $(COMPILE_TIME) $<
 
 clean:
-	rm -rf $(DYLIBNAME) $(STLIBNAME) $(BINS) hiredis-example* *.o *.gcda *.gcno *.gcov
+	rm -rf *.so* $(STLIBNAME) $(BINS) hiredis-example* *.o *.gcda *.gcno *.gcov
 
 dep:
 	$(CC) -MM *.c
@@ -106,7 +111,7 @@ dep:
 install: $(DYLIBNAME) $(STLIBNAME)
 	mkdir -p $(INSTALL_INCLUDE_PATH) $(INSTALL_LIBRARY_PATH)
 	$(INSTALL) hiredis.h async.h adapters $(INSTALL_INCLUDE_PATH)
-	$(INSTALL) $(DYLIBNAME) $(STLIBNAME) $(INSTALL_LIBRARY_PATH)
+	$(INSTALL) *.so* $(STLIBNAME) $(INSTALL_LIBRARY_PATH)
 
 32bit:
 	@echo ""
-- 
